app/src/python/user-scripts/split-fraction-groups.py

```diff
@@ -16,10 +16,16 @@
     print("Loading DICOM file")
     dicom_dataset = pydicom.dcmread(filepath, force=True)
```

Comment:
What do you think about wrapping `pydicom.dcmread()` into a modality specific DICOM reader function? Something like:

```python

MODALITIES = ["RTPLAN", "RTDOSE", "CT", ...]

def read_dicom_by_modality(filepath, modality):
    if not modality in MODALITIES:
        raise ValueError("The supplied modality is not supported") 
    
    ds = pydicom.dcmread(filepath)

    if not ds.Modality == modality:
        raise AttributeError("The supplied DICOM file's modality is not {modality}")

    return ds

Reply:
It would help mitigate the need for `try/except`s like the one in line 19, since the DICOM Standard itself enforces required tags (e.g. all valid RTPLANs have a `FractionGroupSequence`)

Reply:
(Not necessarily in this PR)

Reply:
That seems a little bit like look before you leap...

Reply:
Sort of, but not quite (according to my interpretation of the LBYL vs. EAFP debate). Let me explain:

If you have a function that fails due to an expected error (and failures are low probability), it is more efficient to have:

```python
ds = pydicom.dcmread(filepath)
try:
    rt_plan_specific_function(ds)
except not_an_rt_plan:
    handle_error()
```

vs.:

```python
ds = pydicom.dcmread(filepath)
if is_rt_plan:
    rt_plan_specific_function(ds)
else:
    handle_error()
```

because in the latter case, you are executing "checking code" (a conditional statement) in all scenarios, but in the former, extra code is only executed in (low probability) failure scenarios.

That's the essence of preferring EAFP over LBYL.

However, if you now consider:

```python
ds_plan = load_rt_plan(filepath)
rt_plan_specific_function(ds_plan)
```

vs.

```python
ds = pydicom.dcmread(filepath)
try:
    rt_plan_specific_function(ds)
except not_an_rt_plan:
    handle_error()
```

The former executes extra code, but only *once*, with built in safety and the luxury of being able to avoid the plethora of `try`/`except`s that would be required for *any* RT Plan specific function executed on a generic `ds`. To me, the trade off is worth it in fewer lines of code and greater security.

Also note that if we implement an RT Plan *object* we are essentially doing this anyway...

Reply:
Sure, so is it applicable here? My interpretation of one of the benefits of EAFP is that it partially goes hand in hand with duck typing. It explicitly allows more flexibility. If for example I wanted to write a test where I just tested for this case, I could create just a fraction group and nothing else. I provide to the function only what it needs, and nothing more. And the test raises an exception only if it doesn't have what it needs, and nothing more. So as long as something sufficiently quacks like a duck it is allowed to enter into the function. ... Even if it was a completely different object, even a namedtuple should be allowed in, as long as it has the fraction group attribute.

There are other huge benefits such as avoiding race conditions. But that's not really relevant in this case.

Reply:
I do think though, if we create an RT Plan object, it certainly should have contract with the user. "You supply me something that has these things, and I'll make you an RT plan object."

... I'd probably prefer that an RT plan constructor only required non-duplicate information. I would argue that if fraction group, and all the rest are there, then that information needs to be passed to create the RTPlan object, and to require the other information is to require duplicate information. If conflicting information is provided however such as it looks and quacks like an RTPlan object, but claims to be a dose object, we should raise an error on instantiation. So, in that point, I agree with implementing something "like" this for an RTPlan object. ... Key point is though one should be able to make the object by leaving rendundent information out if they choose. But certainly if redundant information is provided it should raise an error if it is internally inconsistent.

Another piece of rendundent information is CW or CCW etc. The DICOM standard is full of duplicate data.

Reply:
The dicom object itself should always output the full spec however... So if it is created by leaving some items out, the output should fill those in via inference.

Reply:
The `try`/`except` would be moved from the Fraction Group code block to the DICOM reading code block. The potential advantage is if you do other things that are RT Plan specific with that DICOM dataset, you avoid more `try`/`except` blocks. You also avoid issues with *forgetting* `try`/`except` blocks.

I might be wrong, but I don't think duck typing really affects the EAFP vs. LBYL debate. You can code your `if`/`else` with the same scope as your `try`/`except`. I think LBYL is perfectly compatible with duck typing. But your duck typing point does apply to the broader question of whether or not to implement the function, and there I think we have a difference in coding philosophy, which I think amounts to:

- You prefer to keep code as general as possible (in an effort to stay DRY).
- I prefer to keep code more tightly scoped so that all edge cases are handled.

The latter might be more pythonic, but that doesn't necessarily make it the right philosophy for *all* scenarios for PyMedPhys. In my opinion, insisting on keeping functions as general as possible when they are clearly designed for a specific object introduces many unknown unknowns, and a heck of a lot more work/careful thought to ensure known edge cases are handled.

Anyway, let's either discuss in a separate issue, or in person tomorrow :)